src/github.com/matrix-org/dendrite/publicroomsapi/storage/storage.go

```diff
@@ -0,0 +1,211 @@
+	return d.statements.selectPublicRooms(offset, limit, filter)
+}
+
+// UpdateRoomFromEvent updates the database representation of a room from a Matrix event, by
+// checking the event's type to know which attribute to change and using the event's content
+// to define the new value of the attribute.
+// If the event doesn't match with any property used to compute the public room directory,
+// does nothing.
+// If something went wrong during the process, returns an error.
+func (d *PublicRoomsServerDatabase) UpdateRoomFromEvent(event gomatrixserverlib.Event) error {
+	roomID := event.RoomID()
+
+	// Process the event according to its type
+	switch event.Type() {
+	case "m.room.create":
+		return d.statements.insertNewRoom(roomID)
+	case "m.room.member":
+		return d.updateNumJoinedUsers(event, roomID)
+	case "m.room.aliases":
+		return d.updateRoomAliases(event, roomID)
+	case "m.room.canonical_alias":
+		var content common.CanonicalAliasContent
+		field := &(content.Alias)
+		attrName := "canonical_alias"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.name":
+		var content common.NameContent
+		field := &(content.Name)
+		attrName := "name"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.topic":
+		var content common.TopicContent
+		field := &(content.Topic)
+		attrName := "topic"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.avatar":
+		var content common.AvatarContent
+		field := &(content.URL)
+		attrName := "avatar_url"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, "", roomID)
+	case "m.room.history_visibility":
+		var content common.HistoryVisibilityContent
+		field := &(content.HistoryVisibility)
+		attrName := "world_readable"
+		strForTrue := "world_readable"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, strForTrue, roomID)
+	case "m.room.guest_access":
+		var content common.GuestAccessContent
+		field := &(content.GuestAccess)
+		attrName := "guest_can_join"
+		strForTrue := "can_join"
+		return d.updateStringOrBooleanAttribute(attrName, event, &content, field, strForTrue, roomID)
+	}
+
+	// If the event type didn't match, return with no error
+	return nil
+}
+
+// updateNumJoinedUsers updates the number of joined user in the database representation of the room identified by
+// a given room ID, using a given Matrix event.
+// If the event's membership is "join", increments the value, if not, decrements it.
+// Returns an error if the update failed.
+func (d *PublicRoomsServerDatabase) updateNumJoinedUsers(membershipEvent gomatrixserverlib.Event, roomID string) error {
+	membership, err := membershipEvent.Membership()
+	if err != nil {
+		return err
+	}
+
+	if membership == "join" {
+		return d.statements.incrementJoinedMembersInRoom(roomID)
```

Comment:
I think this will break if a user changes their display name since they will send a new join event to the rooms they are already joined to.